tests/js/alias-tests.js

```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2019 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+/* eslint-env es6, node */
+
+'use strict';
+
+var Realm = require('realm');
+var TestCase = require('./asserts');
+var schemas = require('./schemas');
+
+function getRealm() {
+    const schemas = [{
+        name: 'ObjectA',
+        properties: {
+            name: { type: 'string', alias: "otherName"}
+        }
+    }];
+
+    return new Realm({
+        schema: schemas
+    });
+}
+
+function addTestObjects(realm) {
+    realm.beginTransaction();
+    realm.create('ObjectA', {
+        otherName: 'Foo'
+    });
+    realm.create('ObjectA', {
+        otherName: 'Bar'
+    });
+    realm.commitTransaction();
+}
+
+
+module.exports = {
+
+    testAliasInSchema() {
+        const Person = {
+            name: 'Person',
+            properties: {
+                name:     {type: 'string', alias: '_name'},
+                address:  {type: 'string', indexed: true },
+                age:      'double',
+                married:  {type: 'bool', default: false, alias: '_married'},
+                children: {type: 'list', objectType: 'Person', alias: '_children'},
+                parents:  {type: 'linkingObjects', objectType: 'Person', property: 'children', alias: '_parents'},
+            }
+        };
+
+        const realm = new Realm({
+            schema: [Person]
+        });
+
+        const props = realm.schema[0].properties;
+        TestCase.assertEqual(props['name'].alias, '_name');
+        TestCase.assertEqual(props['address'].alias, undefined);
+        TestCase.assertEqual(props['age'].alias, undefined);
+        TestCase.assertEqual(props['married'].alias, '_married');
+        TestCase.assertEqual(props['children'].alias, '_children');
+        TestCase.assertEqual(props['parents'].alias, '_parents');
+    },
+
+
+    testAliasWhenCreatingObjects() {
+        const realm = getRealm();
+        realm.beginTransaction();
+
+        // Creating objects most use the alias
+        realm.create('ObjectA', {
+            otherName: 'Foo'
+        });
+
+        // Creating uses arrays still work
+        realm.create('ObjectA', ['Bar'])
+
+        // Using the internal name instead of the alias throws an exception.
```

Comment:
I don't find this behaviour intuitive. To me an "alias" means another name (ie in addition to but optional). I think the original property should still be valid and useable. At least, I cannot think of why we would want to restrict this.

Reply:
The problem is that if your print the content of an object you will suddenly get 2x the number of properties if we keep both. That feels really award to me. This is how it works in Java (and I believe Cocoa).

Maybe it is the name `alias` that should be changed?

Reply:
What if we still only print the aliased name of a property if one exists, but still allow accessing the property by both the original name and the alias?

Reply:
Yes, I guess we could do that. I'm not 100% sure what the use case would be though?

On the plus side, it grants you the flexibility of a "hidden" API still allowing you to use underlying names. On the negative side, you get an API that seems error prone as mixing "internal" and "public" names probably indicate a user error somewhere?

I'm guessing I just fail to see why you would want to define an alias and still use the internal name? The use case for linking objects where a bit more obvious due to the abstract nature fo defining them.

Reply:
>  I'm not 100% sure what the use case would be though?

The way the tests showcase this feature, the user defines both the name and the alias of a property. This means that the user knows the name in the first place, and may as well refer to the property by its actual name. I might be misunderstanding how you intend to use this feature but here's a made up use case of general aliasing where users may want to use both the alias and the real property name: several classes have properties which have different names but serve a similar purpose, eg: `US_Resident.social_security_number` and `Canadian_Resident.social_insurance_number`. Throughout the code, these properties are referenced via the normal property name, except in a generic function `can_legally_work(resident)` which processes both class types and needs to refer to the property of both in a generic way to check if the id is null. In order to achieve this, both properties have an alias "legal_work_id" applied. In this way the generic code is able to check if the id is null on both class types without having specialised code per class type.

Reply:
The scenario describes the need (or convenience) for general aliasing which I guess wasn't what you had in mind? What is an example of a use case where you would explicitly set up the name and alias of a property but consider it an error if the name was used?

Reply:
I guess I was coming from this from a more statistically typed mindset. The use case I had in mind was e.g renaming fields to conform to naming practises in the language, e.g. the current use case is exposing fields in a subscription in a more idiomatic way. In my case, `createdAt` instead of `created_at` and `timeToLive` instead of `time_to_live`. 

 But I can see the argument for allowing both in a more dynamic language like JS. I'll add support for it.

Reply:
To me, it is most intuitive to be able to use both.

Reply:
After looking at my use case, I don't think that exposing the internal properties are a good idea.

My current use case:

1) Expose `__ResultSets` in a sensible and "safe" manner, including updating fields

Especially updating properties is a problem in Javascript as you cannot override setters. Basically, the pattern is to define a pseudo setter that manipulates a differently named property.

E.g.

```
        const ResultSets = function() {};
        ResultSets.schema = Object.freeze({
            name: '__ResultSets',
            properties: {
                name: { type: 'string', indexed: true, alias: '_name' },
                query: {type: 'string', alias: '_query'},
                matches_property: 'string',
                status: {type: 'int', alias: 'state' },
                error_message: { type: 'string', alias: 'error' },
                query_parse_counter: 'int',
                created_at: { type: 'date', alias: '_createdAt'},
                updated_at: { type: 'date', alias: '_updatedAt'},
                expires_at: { type: 'date', alias: '_expiresAt', optional: true },
                time_to_live: { type: 'int', alias: '_timeToLive', optional: true },
            }
        });
        Object.defineProperty(ResultSets.prototype, "query", {
            enumerable: true,
            get: function() {
                return this._query;
            }
            set: function(val) {
                this._query = validateAndSerializeQuery(val);
                this._updatedAt = new Date();
            }
        });
```

For this to work, I need the following

1) The ability to add a pseudo setter with a name not already used (by using aliases).
2) The ability to query on the pseudo-property. Possible since the internal name and the pseudo-property are the same.

Specifically, this approach means that it is _impossible_ to enumerate the internal names as well since it conflicts with the custom setters/getters. So by exposing both, it is impossible to create a nice wrapper on a managed object. 

Thoughts?

Reply:
That's an interesting approach, but it does seem like you'd need to do some surprising things to get it to work. Wouldn't a more straightforward design be to provide the relevant methods on a subscription object? I don't think js users should be manipulating the `__ResultSets` directly? Ideally they should not know about that implementation and they should be able to get any relevant information from the subscription object?

Reply:
The problem with a standalone subscription object is that it quickly falls apart when it represents a mutating `ResultSets` object. We saw that with the old permission system, where the API just got really weird, quickly.

As long as the subscriptions were static it was a bit more acceptable, but now with the query/timestamps that can change, we would need to do a lot of awkward things to manage subscription object that is immutable.

Just exposing the `__ResultSets`  means we can leverage all the nice things from Realm directly (like changelisteners). The only problem is that we need to mask the underlying implementation as well as add custom logic to setters. Unfortunately, that is rather hard in JS, which is what lead me to the above approach. 

If we supported JS proxies we could have wrapped the `__ResultSets` more nicely, but we don't fully support ES6 across all platforms yet. So, the above approach is as far as I know the best way to hide the underlying implementation.

An alternative could have been to have methods like `updateQuery()` and `updateTtl()`, but that wouldn't stop people from directly manipulating the properties. By renaming the properties to `_XXX` we don't stop them either, but by convention, these are now considered implementation details, and we can then add custom logic to the setters of the JS defined properties.

Reply:
I'm not familiar with the history of the old permission system. I would still favour having `updateQuery()` and `updateTtl()` methods on a subscription object, if it is immutable, it could return an instance to an updated (but immutable) subscription. Maybe I don't fully understand the problems this would create, but it still seems the standard way to operate on subscriptions at a high level. (Also note that in the code sample you posted above, the object is frozen by `Object.freeze` and updates won't be triggered and I think setting the query the way you do will throw a type error?)

Reply:
Note that `Object.freeze()` is only on the schema. Not the rest of the object.

The problem with wrapper objects vs. exposing live objects is three-fold:

1) It goes against everything else we designed Realm to do. Our API for normal model classes exposes "live" objects that are automatically kept up to date. If we believe this is the better approach why don't we do this for our own API as well? 

2) We create a disconnect in our API if some parts of it return static objects while others return "live" objects. It just makes it hard to use and reason about. 

3) Wrapping a live object makes it hard to manage when the underlying properties change. We saw this with the old permission API. Basically, you are forced to add change listeners inside the wrapper and copy out all objects every time something changes to avoid problems if the Realm is closed. You never know if you are looking at the "latest" state and filtering/changelisteners just become hard/impossible. Exposing live objects directly gives us all of this out of the box.

Basically, I don't see the problem in exposing `__ResultSets` directly. We cannot really claim "implementation details" anymore as people have real needs for interacting with it for updating the query or TTL.

Can you elaborate on why you would prefer `updateXXX()` methods on a static Subscription object instead of just exposing the `__ResultSets` ?

Reply:
Will exposing `__ResultSets` make us sensitive to breaking changes in sync?

Reply:
Depending on what type of breaking change you think about. Adding/renaming/removing fields would all be supported using aliases. 

More fundamental changes like removing it completely would obviously not, but we would need to change the wrapper API anyway if that happened.

Reply:
Thanks for explaining, I think I am seeing things more from your perspective now. I do agree that exposing the raw __ResultSets will allow for the most "Realm-like" reactive behaviour. I think the risk is that people can circumvent our intended design, and change things at will like update the query without any real way to guarantee that the new timestamps are updated as well. If you can find a way to make this work with aliases then that sounds like a good way forward.